File: webapi/src/Application/Common/Http/RequestData.cs
namespace NetClock.Application.Common.Http
{
    public class RequestData
    {
        public RequestData()
        {
            TotalItems = 0;
            PageNumber = 1;
            TotalPages = 1;
            PageSize = 10;
            Sorts = string.Empty;
            Filters = string.Empty;
        }

        public int TotalItems { get; set; }

        public int PageNumber { get; set; }

        public int TotalPages { get; set; }

        public int PageSize { get; set; }

        public string Sorts { get; set; }

        public string Filters { get; set; }
    }
}

File: webapp/src/app/models/request-data/request-data.ts
import { LogicalOperator } from './logical-operator';
import { RelationalOperator } from './relational-operator';
import { RequestItemFilter } from './request-item-filter';

export class RequestData<T> {
  hasPreviousPage: boolean;
  hasNextPage: boolean;
  totalItems: number;
  pageNumber: number;
  totalPages: number;
  pageSize: number;
  ratio: number;
  items: T[];
  sorts: string;
  filters: RequestItemFilter[];

  constructor() {
    // @see: webapi/src/Application/Models/Http/RequestData.cs default values.
    this.totalItems = 0;
    this.pageNumber = 1;
    this.totalPages = 1;
    this.pageSize = 10;
    this.ratio = 3;
    this.items = [];
    this.sorts = '';
    this.filters = [];
  }

  addFilter(propertyName: string, operator: RelationalOperator, value: string, concat = LogicalOperator.none): RequestData<T> {
    const filter = new RequestItemFilter(propertyName, operator, value, concat);
    this.filters.push(filter);

    return this;
  }

  removeFilter(filter: RequestItemFilter): RequestData<T> {
    const item = this.filters.indexOf(filter);
    this.filters.splice(item, 1);

    return this;
  }

  cleanFilters(): void {
    this.filters = [];
  }

  stringifyFilters(): string {
    return JSON.stringify(this.filters);
  }

  parseFilters(filters): void {
    this.filters = JSON.parse(filters);
  }
}
